chain/chain/src/flat_storage_resharder.rs

```diff
@@ -0,0 +1,1008 @@
+            split_params;
+        info!(target: "resharding", ?parent_shard, ?left_child_shard, ?right_child_shard, "initiating flat storage shard split");
+        self.check_no_resharding_in_progress()?;
+
+        // Parent shard must be in ready state.
+        let store = self.inner.runtime.store().flat_store();
+        let flat_head = if let FlatStorageStatus::Ready(FlatStorageReadyStatus { flat_head }) =
+            store
+                .get_flat_storage_status(parent_shard)
+                .map_err(|err| Into::<StorageError>::into(err))?
+        {
+            flat_head
+        } else {
+            let err_msg = "flat storage parent shard is not ready!";
+            error!(target: "resharding", ?parent_shard, err_msg);
+            return Err(Error::ReshardingError(err_msg.to_owned()));
+        };
+
+        // Change parent and children shards flat storage status.
+        let mut store_update = store.store_update();
+        let status = SplittingParentStatus {
+            left_child_shard,
+            right_child_shard,
+            shard_layout: shard_layout.clone(),
+            flat_head,
+        };
+        store_update.set_flat_storage_status(
+            parent_shard,
+            FlatStorageStatus::Resharding(FlatStorageReshardingStatus::SplittingParent(
+                status.clone(),
+            )),
+        );
+        store_update.set_flat_storage_status(
+            left_child_shard,
+            FlatStorageStatus::Resharding(FlatStorageReshardingStatus::CreatingChild),
+        );
+        store_update.set_flat_storage_status(
+            right_child_shard,
+            FlatStorageStatus::Resharding(FlatStorageReshardingStatus::CreatingChild),
+        );
+        store_update.commit()?;
+
+        self.schedule_split_shard(parent_shard, &status, scheduler, controller);
+        Ok(())
+    }
+
+    /// Returns an error if a resharding event is in progress.
+    fn check_no_resharding_in_progress(&self) -> Result<(), StorageError> {
+        // Do not allow multiple resharding events in parallel.
+        if self.resharding_event().is_some() {
+            error!(target: "resharding", "trying to start a new flat storage resharding event while one is already in progress!");
+            Err(StorageError::FlatStorageReshardingAlreadyInProgress)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn set_resharding_event(&self, event: FlatStorageReshardingEventStatus) {
+        *self.inner.resharding_event.lock().unwrap() = Some(event);
+    }
+
+    /// Returns the current in-progress resharding event, if any.
+    pub fn resharding_event(&self) -> Option<FlatStorageReshardingEventStatus> {
+        self.inner.resharding_event.lock().unwrap().clone()
+    }
+
+    /// Schedules a task to split a shard.
+    fn schedule_split_shard(
+        &self,
+        parent_shard: ShardUId,
+        status: &SplittingParentStatus,
+        scheduler: &dyn FlatStorageResharderScheduler,
```

Comment:
nothing seems to implement FlatStorageResharderScheduler right now?

Reply:
Yes for now it's used only in test

I could make this into a sender or have the actor impl FlatStorageResharderScheduler 🤔